README.md
# bar_chart_race

A library to draw fantastic bar charts race in Flutter  
<img src="https://github.com/younes38/bar_chart_race/blob/main/bar_chart_race.gif" width="220" height="400" />

## Usage

### Let's get started

add the dependencies to your app:

```yaml
dependencies:
  bar_chart_race: ^0.0.4
```

then import it:  
`import 'package:bar_chart_race/bar_chart_race.dart';`

then you can simply use:

```Dart
BarChartRace(
            data: data,
            initialPlayState: true,
          )
```

this will create a bar chart with random colors and animate it.  
the `data` and `initialPlayState` are required and here the exmplanation for most of the properties

- `data`: the data you are going to show and it should contains at lease two rows and two columns
- `initialPlayState`: if it's `true` then the bar chart will be animated
- `framesPerSecond`: defines the number of frames to show per seconds
- `columnsLabel`: represents the name of the columns
- `statesLabel`: represents the name of the rows (usually time)
- `numberOfRactanglesToShow`: represents the number of the first columns to show
- `title`: The title of the bar chart race
- `columnsColor`: the color of each rectangle

## Contribution

pull requests are welcome!

## Last but not least

the package is still in its first version so it may contains some issues.


CHANGELOG.md
## 0.0.4 - description updated

## 0.0.3 - Full example added

## 0.0.2 - enhanced readme and fixes some issues

## 0.0.1 - Released on (2021 May 2)


test/bar_chart_race_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:bar_chart_race/bar_chart_race.dart';

void main() {}


lib/bar_chart_race.dart
library bar_chart_race;

import "package:flutter/material.dart";
import 'dart:math' as math;
import 'models/rectangle.dart';
import 'paint/my_state_paint.dart';

/// Creates an interactive bar chart depending on the provided [data], which looks like a race
///
/// [data] should be a matrix of double with at least two columns and two rows, and stores data in cumulative way
/// [initialPlayState] if true then the bar chart will be animated, else it will show the first row of data and stop
/// [framesPerSecond] defines the number of frames to show per seconds
/// [columnsLabel] represents the name of the columns
/// [statesLabel] represents the name of the rows (usually time)
/// [numberOfRactanglesToShow] represents the number of the first columns to show
/// [title] title of the bar chart race
/// [columnsColor] the color of each column
class BarChartRace extends StatefulWidget {
  /// each row represents a state and each column represents one instance
  /// data should contains at least two rows and two columns, else it's not logic
  /// data should be cumulative
  final List<List<double>> data;

  /// if it's true then the bar chart will be animated
  final bool initialPlayState;

  /// number of frame to show in one second
  final double framesPerSecond;

  /// represent the number of frames to show between two states (two consecutive rows)
  final int framesBetweenTwoStates;

  /// a list of labels for each column.
  /// length of the this list show be equals to the number of columns of [data].
  ///
  /// for examle if you are going to study countries provides the list of countries with the same order as the columns in data.
  final List<String> columnsLabel;

  /// one label for each state.
  /// if your state is the time you need to prodive the time of each state
  /// length of [stateLabels] list should be equal to the number of rows in [data].
  final List<String> statesLabel;

  /// number of rectangles to show in the UI.
  ///
  /// For example if you are studying ten countries, you can show the first five countries only
  /// by default it equals to the number of columns in the data
  final int numberOfRactanglesToShow;

  /// a color for each column,
  ///
  /// for example if you are studying countries, you can assign to each country a color
  /// colors should be ordred as the column in the [data]
  /// by default it uses random colors
  final List<Color> columnsColor;

  /// represent the title of the chart
  final String title;

  /// the styling for the text
  final TextStyle titleTextStyle;

  /// the height of the rectangle
  final double rectangleHeight;

  const BarChartRace({
    Key key,
    @required this.data,
    @required this.initialPlayState,
    this.framesPerSecond = 20,
    this.framesBetweenTwoStates = 40,
    this.rectangleHeight = 27,
    this.numberOfRactanglesToShow = 5,
    this.columnsLabel,
    this.statesLabel,
    this.columnsColor,
    this.title,
    this.titleTextStyle,
  }) : super(key: key);

  @override
  _BarChartRaceState createState() => _BarChartRaceState();
}

class _BarChartRaceState extends State<BarChartRace> {
  int nbStates;
  int nbParticipants;
  // data of preprocessing the initial data
  List<List<Rectangle>> preparedData;
  // current data to show
  List<Rectangle> currentData;

  @override
  void initState() {
    // init local variables
    nbStates = widget.data.length;
    nbParticipants = widget.data[0].length;
    // prepare data to be shown in the customPainer
    preparedData = prepareData(widget.data);
    currentData = preparedData[0];
    if (widget.initialPlayState) play();
    super.initState();
  }

  // re-build
  void _update() {
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (_, constraints) => CustomPaint(
        painter: MyStatePaint(
          currentState: currentData,
          numberOfRactanglesToShow: widget.numberOfRactanglesToShow,
          rectHeight: widget.rectangleHeight,
          maxValue: currentData[0].maxValue,
          totalWidth: constraints.maxWidth * 0.75,
          title: widget.title,
          titleTextStyle: widget.titleTextStyle,
        ),
      ),
    );
  }

  void play() async {
    for (int i = 1; i < nbStates; i++) {
      await makeTransition(preparedData[i - 1], preparedData[i]);
    }
  }

  Future<void> makeTransition(
      List<Rectangle> before, List<Rectangle> after) async {
    int nbFrames = widget.framesBetweenTwoStates;
    int fps = widget.framesBetweenTwoStates;

    for (int k = 1; k <= nbFrames; k++) {
      // for each frame we update the current value
      for (int i = 0; i < nbParticipants; i++) {
        // get the difference between two states
        double posDiff = (after[i].position - before[i].position) / nbFrames;
        double lengthDiff = (after[i].length - before[i].length) / nbFrames;
        double valueDiff = (after[i].value - before[i].value) / nbFrames;
        double maxValueDiff =
            (after[i].maxValue - before[i].maxValue) / nbFrames;
        // add the new differences
        currentData[i].length = before[i].length + lengthDiff * k;
        currentData[i].position = before[i].position + posDiff * k;
        currentData[i].value = before[i].value + valueDiff * k;
        currentData[i].maxValue = before[i].maxValue + maxValueDiff * k;
        // upadte the labels
        if ((widget.columnsLabel?.length ?? 0) > 0) {
          currentData[i].label = widget.columnsLabel[i];
        }
        if ((widget.statesLabel?.length ?? 0) > 0) {
          currentData[i].stateLabel = before[i].stateLabel;
        }
      }
      // rebuild the UI
      _update();
      await Future.delayed(Duration(milliseconds: 1000 ~/ fps));
    }
  }

  // prepare data so that it can be shown,
  List<List<Rectangle>> prepareData(List<List<double>> data) {
    List<List<Rectangle>> resultData = [];
    // for each state (a row from data), we sort row without modifying by using anothe list of indexes
    for (int i = 0; i < nbStates; i++) {
      List<int> indexes = List.generate(nbParticipants, (index) => index);
      // sort the indexes in deceasing order based on the data in the row
      indexes.sort((int a, int b) {
        return data[i][b].compareTo(data[i][a]);
      });
      // get the max value, which is in the first index
      double maxValue = data[i][indexes[0]];
      List<Rectangle> currentState = List(nbParticipants);
      for (int j = 0; j < nbParticipants; j++) {
        int index = indexes[j];
        // generate a random color to use in case the colors are not provided
        Color randomColor =
            Color((math.Random().nextDouble() * 0xFFFFFF).toInt())
                .withOpacity(1.0);
        // create the Rectable object which will be used to draw the chart
        currentState[index] = Rectangle(
          maxValue: maxValue,
          length: data[i][index] / maxValue,
          position: 1.0 * j,
          value: data[i][index],
          color: widget.columnsColor == null
              ? randomColor
              : widget.columnsColor[index],
          stateLabel: widget.statesLabel != null ? widget.statesLabel[i] : null,
          label:
              widget.columnsLabel != null ? widget.columnsLabel[index] : null,
        );
      }
      resultData.add(currentState);
    }
    return resultData;
  }
}


lib/paint/my_state_paint.dart
import 'package:bar_chart_race/models/rectangle.dart';
import 'package:flutter/material.dart';

class MyStatePaint extends CustomPainter {
  final List<Rectangle> currentState;
  final double maxValue;
  final double rectHeight;
  final double totalWidth;
  final int numberOfRactanglesToShow;
  final String title;
  final TextStyle titleTextStyle;

  // rectangle paint
  final Paint rectPaint;
  // lines paint
  final Paint linePaint;

  // the lenth of the maximum value
  double maxLength;
  MyStatePaint({
    @required this.currentState,
    @required this.maxValue,
    @required this.numberOfRactanglesToShow,
    @required this.totalWidth,
    @required this.rectHeight,
    this.title,
    this.titleTextStyle,
  })  : rectPaint = Paint(),
        linePaint = Paint() {
    // proporties for rectangle paint
    rectPaint.style = PaintingStyle.fill;
    rectPaint.strokeWidth = 0;
    rectPaint.strokeCap = StrokeCap.round;

    // proporties for lines paint
    linePaint.color = Colors.grey;
    linePaint.style = PaintingStyle.stroke;
    linePaint.strokeWidth = 2;
    linePaint.strokeCap = StrokeCap.round;

    maxLength = totalWidth * 0.9;
  }

  final double spaceBetweenTwoRectangles = 6;
  final double yShift = 50;
  final double xShift = 70;
  // define text painter to paint text (write text)
  final TextPainter textPainter = TextPainter(
    textAlign: TextAlign.center,
    textDirection: TextDirection.ltr,
  );
  final TextStyle textStyle = TextStyle(
    color: Colors.black,
    fontSize: 15.0,
  );

  // define text painter to paint title
  final TextPainter titlePainter = TextPainter(
    textAlign: TextAlign.center,
    textDirection: TextDirection.ltr,
  );
  final TextStyle textStyleBold = TextStyle(
    color: Colors.black,
    fontSize: 20.0,
    fontWeight: FontWeight.bold,
  );

  @override
  void paint(Canvas canvas, Size size) {
    canvas.translate(xShift, yShift);
    // draw title if not null
    if (title != null) _drawTitle(canvas, title);
    // first we draw the line under the ractangles
    _drawLines(canvas);
    // we draw the rectangles
    for (int i = 0; i < currentState.length; i++) {
      _drawRectangle(currentState[i], canvas);
    }
    // draw current state label
    String stateLabel = currentState[0].stateLabel;

    if (stateLabel != null) {
      _drawStateLabel(canvas, size, stateLabel);
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }

  // TODO test this
  void _drawStateLabel(Canvas canvas, Size size, String label) {
    // draw value (text)
    textPainter.text = TextSpan(
      text: label,
      style: textStyleBold,
    );
    double x = totalWidth;
    double y = rectHeight * numberOfRactanglesToShow;
    print(y);
    canvas.save();
    textPainter.layout();
    canvas.translate(x, y);
    textPainter.paint(
      canvas,
      Offset(
        -textPainter.width,
        textPainter.height,
      ),
    );
    canvas.restore();
  }

  void _drawTitle(Canvas canvas, String title) {
    // draw the title of the chart
    textPainter.text = TextSpan(
      text: title,
      style: titleTextStyle ?? textStyleBold,
    );
    double x = totalWidth / 2;
    double y = -40;
    canvas.save();
    textPainter.layout();
    canvas.translate(x, y + 3);
    textPainter.paint(
      canvas,
      Offset(
        -textPainter.width / 2,
        -textPainter.height / 2,
      ),
    );
    canvas.restore();
  }

  void _drawRectangle(Rectangle rect, Canvas canvas) {
    // draw rectangle
    Path path = Path();
    double maxHeight =
        numberOfRactanglesToShow * (rectHeight + spaceBetweenTwoRectangles) -
            spaceBetweenTwoRectangles;
    // define postitons of the four corner to draw the rectangle
    double x1 = 0,
        y1 = rect.position * (rectHeight + spaceBetweenTwoRectangles);
    // if the rectancles if outside, we don't draw it
    if (y1 >= maxHeight) return;
    // min is to draw a rectangle partially, (in case it's showning up or hiding)
    double x2 = rect.length * maxLength, y2 = min(y1 + rectHeight, maxHeight);

    path.moveTo(x1, y1);
    path.lineTo(x2, y1);
    path.lineTo(x2, y2);
    path.lineTo(x1, y2);
    rectPaint.color = rect.color;
    canvas.drawPath(path, rectPaint);

    // draw value (text)
    String value = rect.value.round().toString();
    if (value.length > 5) {
      value = value.substring(0, 5) + "..";
    }
    textPainter.text = TextSpan(
      text: value,
      style: textStyle,
    );
    canvas.save();
    textPainter.layout();
    canvas.translate(x2, y1 + 3);
    textPainter.paint(
      canvas,
      Offset(
        5,
        0,
      ),
    );
    canvas.restore();

    // draw the title for each rectangle
    String label = rect.label;
    if (label.length > 7) {
      label = label.substring(0, 7) + "..";
    }
    textPainter.text = TextSpan(
      text: label,
      style: textStyle,
    );
    canvas.save();
    textPainter.layout();
    canvas.translate(0, y1 + 3);
    textPainter.paint(
      canvas,
      Offset(
        -textPainter.width - 2,
        0,
      ),
    );
    canvas.restore();
  }

  // draw the lines with the respective value based on the current maximum value
  void _drawLines(Canvas canvas) {
    double lastDigit = maxValue;
    double p = 1;
    while (lastDigit >= 10) {
      lastDigit = lastDigit / 10.0;
      p *= 10;
    }
    double step;
    if (lastDigit < 3) {
      step = 0.5 * p / maxValue * maxLength;
    } else {
      step = 2 * p / maxValue * maxLength;
    }
    double posX = step;
    while (posX <= totalWidth) {
      double value = posX / maxLength * maxValue;
      _drawLine(canvas, posX, value.round());
      posX += step;
    }
  }

  void _drawLine(Canvas canvas, double posX, int value) {
    Path path = Path();
    // define the two point of the line
    double x1 = posX, y1 = 0;
    double x2 = posX,
        y2 = numberOfRactanglesToShow *
                (rectHeight + spaceBetweenTwoRectangles) -
            spaceBetweenTwoRectangles;
    path.moveTo(x1, y1);
    path.lineTo(x2, y2);
    canvas.drawPath(path, linePaint);
    // draw the value (text)
    textPainter.text = TextSpan(
      text: value.round().toString(),
      style: textStyle,
    );
    canvas.save();
    textPainter.layout();
    canvas.translate(x1, y1);
    textPainter.paint(
      canvas,
      Offset(-(textPainter.width / 2), -(textPainter.height)),
    );
    canvas.restore();
  }

  double min(double a, double b) {
    return a < b ? a : b;
  }
}


example/lib/random_data_example.dart
import 'package:bar_chart_race/bar_chart_race.dart';
import 'package:flutter/material.dart';
import 'dart:math' as math;

class RandomDataExample extends StatefulWidget {
  @override
  _RandomDataExampleState createState() => _RandomDataExampleState();
}

class _RandomDataExampleState extends State<RandomDataExample> {
  List<List<double>> data;
  bool isPlaying = false;
  @override
  void initState() {
    data = generateGoodRandomData(20, 5);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Bar chart race"),
        backgroundColor: Colors.brown,
        elevation: 0,
      ),
      body: Column(
        children: [
          SizedBox(
            height: 150,
          ),
          BarChartRace(
            data: data,
            initialPlayState: true,
            columnsColor: [
              Color(0xFFFF9900),
              Color(0xFF4285F4),
              Color(0xFFA2AAAD),
              Color(0xFFF40009),
              Color(0xFF212326),
            ],
            framesPerSecond: 60,
            framesBetweenTwoStates: 30,
            numberOfRactanglesToShow: 4,
            title: "Top companies revenue",
            columnsLabel: ["Amazon", "Google", "Apple", "Coca", "Huawei"],
            statesLabel: List.generate(
              30,
              (index) => formatDate(
                DateTime.now().add(
                  Duration(days: index),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  List<List<double>> generateGoodRandomData(int nbRows, int nbColumns) {
    List<List<double>> data = List.generate(nbRows, (index) => List());
    for (int j = 0; j < nbColumns; j++) {
      data[0].add(j * 10.0);
    }
    for (int i = 1; i < nbRows; i++) {
      for (int j = 0; j < nbColumns; j++) {
        data[i].add(data[i - 1][j] +
            (nbColumns - j) +
            math.Random().nextDouble() * 20 +
            (j == 2 ? 10 : 0));
      }
    }
    return data;
  }

  formatDate(DateTime date) {
    int day = date.day;
    int month = date.month;
    int year = date.year;
    return "${months[month - 1]} $day, $year";
  }
}

List<String> months = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

List<String> weekDays = [
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
  "Sunday",
];


example/lib/main.dart
import 'package:bar_chart_race/bar_chart_race.dart';
import 'package:flutter/material.dart';
import 'dart:math' as math;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        // is not restarted.
        primarySwatch: Colors.blue,

        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: RandomDataExample(),
    );
  }
}

class RandomDataExample extends StatefulWidget {
  @override
  _RandomDataExampleState createState() => _RandomDataExampleState();
}

class _RandomDataExampleState extends State<RandomDataExample> {
  List<List<double>> data;
  bool isPlaying = false;
  @override
  void initState() {
    data = generateGoodRandomData(20, 5);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Bar chart race"),
        backgroundColor: Colors.brown,
        elevation: 0,
      ),
      body: Column(
        children: [
          SizedBox(
            height: 150,
          ),
          BarChartRace(
            data: data,
            initialPlayState: true,
            columnsColor: [
              Color(0xFFFF9900),
              Color(0xFF4285F4),
              Color(0xFFA2AAAD),
              Color(0xFFF40009),
              Color(0xFF212326),
            ],
            framesPerSecond: 60,
            framesBetweenTwoStates: 30,
            numberOfRactanglesToShow: 4,
            title: "Top companies revenue",
            columnsLabel: ["Amazon", "Google", "Apple", "Coca", "Huawei"],
            statesLabel: List.generate(
              30,
              (index) => formatDate(
                DateTime.now().add(
                  Duration(days: index),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  List<List<double>> generateGoodRandomData(int nbRows, int nbColumns) {
    List<List<double>> data = List.generate(nbRows, (index) => List());
    for (int j = 0; j < nbColumns; j++) {
      data[0].add(j * 10.0);
    }
    for (int i = 1; i < nbRows; i++) {
      for (int j = 0; j < nbColumns; j++) {
        data[i].add(data[i - 1][j] +
            (nbColumns - j) +
            math.Random().nextDouble() * 20 +
            (j == 2 ? 10 : 0));
      }
    }
    return data;
  }

  formatDate(DateTime date) {
    int day = date.day;
    int month = date.month;
    int year = date.year;
    return "${months[month - 1]} $day, $year";
  }
}

List<String> months = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];

List<String> weekDays = [
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
  "Sunday",
];
